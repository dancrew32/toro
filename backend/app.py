import json

from flask import Flask
from flask import jsonify

app = Flask(__name__)


def serve_json(path):
    with open(path, 'rb') as f:
        data = json.load(f)
    response = jsonify(data)
    response.headers.add('Access-Control-Allow-Origin', '*')
    return response


@app.route('/tables')
def tables():
    return serve_json('./data/tables.json')


@app.route('/metrics/<int:table_id>')
def metrics(table_id: int):
    return serve_json(f'./data/metrics.{table_id}.json')
