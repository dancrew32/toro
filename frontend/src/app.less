@import (reference) "./colors.less";

@navHeight: 40px;

// Reset and full screen requiement
html,
body {
  margin: 0;
  padding: 0;
  height: 100%;
}

// universal box sizing
*,
*:before,
*:after {
  box-sizing: border-box;
}

// style this up
body {
  font-family: "DM Mono", monospace;
  font-size: 13px;
  background-color: @toro_dark;
  color: @toro_light;
}

// React app root (full height - nav)
#root {
  height: 100%;
  padding-top: @navHeight;
}

// Body links
a {
  text-decoration: none;
  color: @toro_light;
  &:hover {
    text-decoration: underline;
    color: darken(@toro_light, 10%);
  }
  &.active,
  &:active {
    color: @toro_primary;
  }
}

// headings
h1,
h2,
h3 {
  margin: 0 0 10px;
  padding: 0;
  font-weight: 500;
}

code {
  font-size: 12px;
  color: @toro_primary;
}

nav {
  background-color: @toro_light;
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  min-height: @navHeight;
  border-bottom: 1px solid @toro_mid;
  a {
    font-size: 16px;
    font-weight: 500;
    text-decoration: none;
    color: @toro_dark;
    border: 1px solid @toro_light;
    padding: 10px 20px;
    display: inline-block;
    &:hover {
      color: lighten(@toro_dark, 10%);
    }
    &:active {
      color: @toro_primary;
    }
  }
}

// Table defaults
table {
  margin: 0 0 3px;
  padding: 0;
  border-collapse: collapse;
  border-spacing: 0;
  background-image: none;
  width: 100%;
  max-width: 550px;
}
th,
td {
  text-align: left;
  border: 1px solid @toro_mid;
  padding: 3px;
}

// table list view
aside {
  flex: 0 0 220px;
  border-right: 1px solid @toro_mid;
  height: 100%;
  overflow-x: auto;
  overflow-y: scroll;
}

// main column (metrics)
main {
  flex: 1;
  height: 100%;
  overflow-y: scroll;
}

// homepage article layout
article {
  margin: 0 auto;
  max-width: 550px;
}

// full height container, and safe padding .inner
.container {
  display: flex;
  height: 100%;
}
.inner {
  padding: 10px;
}

// search input
.search {
  position: relative;
  input {
    display: block;
    margin: 0 0 10px;
    padding: 5px 10px;
    width: 100%;
    border: 1px solid @toro_mid;
    border-radius: 3px;
    &:focus {
      border-color: @toro_primary;
    }
  }
  .clear-button {
    position: absolute;
    right: 0;
    top: 3px;
    font-size: 14px;
    font-weight: 500;
    background-color: transparent;
    border: none;
    cursor: pointer;
    overflow: hidden;
  }
}

// table list view
.table-schema {
  position: relative;
  a {
    padding: 3px 0;
    display: block;
  }
}
.table-schema-active {
  a {
    color: @toro_primary;
  }
}

// expandable schema.table columns
.table-columns {
  overflow: hidden;
  transition: max-height 0.2s ease-out;
  height: auto;
  max-height: 600px;
  margin-left: 10px;
}
.table-columns-collapsed {
  max-height: 0;
}
.table-columns-toggle {
  background-color: @toro_dark;
  position: absolute;
  top: 0;
  right: -5px;
  cursor: pointer;
  display: inline-block;
  padding: 3px 5px;
  width: 25px;
  text-align: center;
  transition: transform 0.1s ease;
  transform: rotate(90deg);
  user-select: none;
  opacity: 0.8;
  &:hover {
    opacity: 1;
  }
}
.table-columns-toggle-open {
  transform: rotate(0deg);
}
