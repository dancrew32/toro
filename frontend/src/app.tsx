import * as React from "react";
import * as ReactDOM from "react-dom";
import { Router } from "./router";
import { AppProvider } from "./store";
import "./app.less";

export function App() {
  return (
    <AppProvider>
      <Router />
    </AppProvider>
  );
}

const root = document.getElementById("root") || document.createElement("div");
ReactDOM.render(<App />, root);
