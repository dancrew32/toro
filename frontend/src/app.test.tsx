import * as React from "react";
import * as TestRenderer from "react-test-renderer";

import { App } from "./app";

test("<App />", () => {
  const tree = TestRenderer.create(<App />).toJSON();
  expect(tree).toMatchSnapshot();
});
