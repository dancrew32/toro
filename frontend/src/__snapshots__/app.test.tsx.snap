// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> 1`] = `
<div
  className="container"
>
  <nav>
    <a
      className="active"
      href="/"
      onClick={[Function]}
    >
      Toro
    </a>
    <a
      className=""
      href="/tables"
      onClick={[Function]}
    >
      Tables
    </a>
  </nav>
  <article>
    <div
      className="inner"
    >
      <h2>
        Overview
      </h2>
      <p>
        Here at Toro we have to build tools that make navigating and managing a large number of disparate metrics and datasets fast and approachable to a variety of different end users - analysts, data scientists, PMs, and engineers.
      </p>
      <p>
        For this challenge, you will be given APIs to retrieve tables and metrics associated with them and your goal is to create an interface for users to navigate these metrics.
      </p>
      <h2>
        Requirements
      </h2>
      <p>
        There are no restrictions on how to design and implement the interface, but be ready to discuss both your technical and product decisions. We expect you to spend no more than two hours on this project, and if you have any questions or issues, feel free to reach out to [redacted].
      </p>
      <p>
        When youâ€™re done, email [redacted] with a link to a GitHub repo or a zip of your files, and instructions on how to run your project.
      </p>
      <p>
        For the final deliverable, as a user of your interface, I should: Be able to navigate to all of the metrics available to me See all of the configuration for a particular metric
      </p>
      <h2>
        APIs
      </h2>
      <p>
        List of all tables https://interview.torodata.io/tables
      </p>
      <p>
        List of metrics for a given table
        https://interview.torodata.io/metrics/{tableId} 
        e.g - all of the metrics for the table with an id=1 would be at https://interview.torodata.io/metrics/1
      </p>
    </div>
  </article>
</div>
`;
